lib/puppet/application/lookup.rb

```diff
@@ -49,8 +60,49 @@ class Puppet::Application::Lookup < Puppet::Application
   def run_command
     options[:keys] = command_line.args
 
-    puts "Hello World!"
-    puts options
+    if options[:keys].empty?
+     raise "No keys were given to lookup."
+    end
+
+    if !options[:node]
+      raise "No node was given via the '--node' flag for the scope of the lookup."
+    end
+
+    if (options[:sort_merge_arrays] || options[:merge_hash_arrays] || options[:prefix] || options[:unpack_arrays]) && options[:merge] != 'deep'
+      raise "The options #{DEEP_MERGE_OPTIONS} are only available with '--merge deep'\n#{RUNHELP}"
+    end
+
+    scope = generate_scope
+
+    use_default_value = !options[:default_value].nil?
+    merge_options = nil
+
+    if options[:merge]
+      if options[:merge] == 'deep'
+        merge_options = {'strategy' => 'deep',
+          'sort_merge_arrays' => !options[:sort_merge_arrays].nil?,
+          'merge_hash_arrays' => !options[:merge_hash_arrays].nil?}
+
+        if options[:prefix]
+          merge_options.merge({'prefix' => options[:prefix]})
+        end
+
+        if options[:unpack_arrays]
+          merge_options.merge({'unpack_arrays' => options[:unpack_arrays]})
+        end
+
+      else
+        merge_options = {'strategy' => options[:merge]}
+      end
+    end
+
+    puts Puppet::Pops::Lookup.lookup(scope, options[:keys], options[:type], options[:default_value], use_default_value, {}, {}, merge_options)
   end
 
+  def generate_scope
+    node = Puppet::Node.indirection.find("#{options[:node]}")
+    compiler = Puppet::Parser::Compiler.new(node)
+    compiler.compile
```

Comment:
I don't think that is very difficult - there are very concrete steps taken in the "compile" method. I imagine that the faster way would only do the "initial import" - that would only do the evaluation of the site.pp top scope, it would load/parse all included classes and register them, but would not completely evaluate everything (thus, it would not drill down into recursive includes). We can probably also prevent the initial-import from taking place, which means there is a top scope with only facts (node parameters), and settings (in the settings scope) available. With both these setups, the infrastructure (binder, loaders, compiler, top-scope, etc.) are all initialized and possible to use - i.e. it is possible to use puppet functions for modules in data etc.

The same effect is obtained by doing a compile with an empty site.pp; this is perhaps a more appealing option - simply change the "manifest" setting for the compile. Remember that in essence, the lookup command is a shortcut equivalence to writing a .pp file with a lookup call in it and then using apply.